### ISSUE TITLE
Using cog3pio to determine byte ranges in COG files?

### ISSUE THREAD

**Issue by TomNicholas** [weiji14/cog3pio#16]
Basically same question as https://github.com/gauteh/hidefix/issues/38#issue-2236324824 but for this library 😁 

I'm building [VirtualiZarr](https://github.com/TomNicholas/VirtualiZarr), an evolution of [kerchunk](https://github.com/fsspec/kerchunk), that allows you to determine byte ranges of chunks in netCDF files, but then concatenate the virtual representation of those chunks using xarray's API.

This works by creating a `ChunkManifest` object in-memory (one per netCDF Variable per file initially), then defining ways to merge those manifests.

What I'm wondering is if [cog3pio](https://github.com/weiji14/cog3pio)'s code could be useful to me as a way to generate the `ChunkManifest` for a netCDF file without using `kerchunk`/`fsspec` (see [this issue](https://github.com/TomNicholas/VirtualiZarr/issues/78)). In other words I _use cog3pio only to determine the byte ranges_, not for actually reading the data. (I plan to actually read the bytes later as if it were zarr using the rust `object-store` crate, see https://github.com/zarr-developers/zarr-python/pull/1661).

Q's:
- Is this idea dumb?
- Does cog3pio expose the byte range information currently?
- Can cog3pio read over S3?

cc @norlandrhagen

**Comment by weiji14** [comment]
Oh hi Tom, I was just chatting to @norlandrhagen at the Pangeo weekly meeting :laughing:. I will say that yes, it should be possible to work out the byte ranges from just the GeoTIFF's header, and we can expose that via an API function somehow. I just need to figure out how GDAL does this, and re-implement it here (easier said than done).

Note that I'm already using `object-store` in cog3pio (https://github.com/weiji14/cog3pio/pull/5), and passing a HTTP url to a GeoTIFF should already work. Reading from s3 (or azure, gcp, etc) will work too if I enable the feature flag here and recompile:

https://github.com/weiji14/cog3pio/blob/33285da1280f165d623457c73c1ae55c1674472e/Cargo.toml#L17

I'm aware that the Zarr v3 implementation is using object-store, and keeping an eye on progress at https://github.com/roeap/object-store-python. Would definitely be keen to standardize on object-store as the 'fsspec-for-Rust'.

**Comment by TomNicholas** [comment]
Awesome! Thanks @weiji14 

> Would definitely be keen to standardize on object-store as the 'fsspec-for-Rust'.

Yeah this would be great, and I like the way you've described the aim there.

**Comment by martindurant** [comment]
I would add, that there's no reason that kerchunk needs to use fsspec - it's really a package of ways to make reference files. Therefore, if you come up with a way to get COG offsets, it can easily live there together with the other reference makers. Does [TIFFFile](https://github.com/cgohlke/tifffile/blob/master/tifffile/tifffile.py#L12921) do  the required work too?

**Comment by weiji14** [comment]
Thanks for chiming in @print-sid8! It's on the roadmap to port the backend to `async-tiff`, though it's lower on my priority list for 2025 right now 🙈

> This can understand the headers of a COG, then decide which byte ranges to read for a given geometry AOI.
> 
> maybe its useful for you to port to rust?

Yes, I think it would be useful to port the logic in your [`compute_tile_indices`](https://github.com/terrafloww/rasteret/blob/585e65b97f127570394bb6d41d30ab94e4dec7ca/src/rasteret/fetch/cog.py#L196-L284) function (+some other ones) to Rust. Just need to find the time for it 😭 

> async-tiff by Kyle Barron already has most of what I made, he released his rust codebase after I made mine, I would prefer to offload the high concurrent reading bits from my httpx implementation to his rust implementation.

I see you've switched to `aiohttp` already in https://github.com/terrafloww/rasteret/pull/8 😄 My hunch is that you've probably got most of the speedup gain already by switching the network call from a synchronous to asynchronous I/O.

**Comment by print-sid8** [comment]
> I see you've switched to aiohttp already in https://github.com/terrafloww/rasteret/pull/8 😄 My hunch is that you've probably got most of the speedup gain already by switching the network call from a synchronous to asynchronous I/O.

Oh yes, jeff's benchmarks obviously made me move to aiohttp, and it has had very good improvements on certain parts of my library, like the bootstrapping of the COG header cache. So similar to you, i don't think ill spend to much time trying to refactor my codebase again to add async-tiff immediately to rasteret.

I would rather take time to understand efforts like cog3pio and contribute what I can.

> Yes, I think it would be useful to port the logic in your [compute_tile_indices](https://github.com/terrafloww/rasteret/blob/585e65b97f127570394bb6d41d30ab94e4dec7ca/src/rasteret/fetch/cog.py#L196-L284) function (+some other ones) to Rust. Just need to find the time for it 😭 

I'm no rust sensei but I can try. 😄 Let me know your ideas and philosophy behind where such similar functions might help in cog3pio, I'll try and understand and give it a shot.

**Comment by weiji14** [comment]
> I would rather take time to understand efforts like cog3pio and contribute what I can.
> 
> > Yes, I think it would be useful to port the logic in your [compute_tile_indices](https://github.com/terrafloww/rasteret/blob/585e65b97f127570394bb6d41d30ab94e4dec7ca/src/rasteret/fetch/cog.py#L196-L284) function (+some other ones) to Rust. Just need to find the time for it 😭
> 
> I'm no rust sensei but I can try. 😄 Let me know your ideas and philosophy behind where such similar functions might help in cog3pio, I'll try and understand and give it a shot.

Cool, would be great to get you into Rust 😄 My long-term idea is to have the following dependency chain:

`async-tiff` (Rust 🦀 core, low-level) -> `geotiff` (Rust 🦀 'geo' 🌐 lib, medium-level) -> `cog3pio` (Rust+Python 🦀🐍 high-level bindings)

Right now, the picture is a little complicated, as there isn't a clear separation of pure-TIFF/GeoTIFF and low-level/high-level API interfaces. Again, limiting factor is just having the time + people to do all this!

If you just want to focus on porting the 'compute_tile_indices' function, I'd recommend checking out this issue - https://github.com/georust/geotiff/issues/23, and we can continue the discussion there.

**Comment by weiji14** [comment]
> So for a VirtualiZarr (geo)tiff parser, which of those layers is needed? All 3? (For a parser written in python)

Yeah, about that... at the moment, `async-tiff` has some geo-bits which [`virtual-tiff`](https://github.com/virtual-zarr/virtual-tiff) is using, which is then used by Virtualizarr. It depends on what Virtualizarr is after, if you want to virtualize any TIFFs (including non-geo ones), then this current `async-tiff` -> `virtual-tiff` -> `Virtualizarr` setup probably works fine. If Virtualizarr just wants to focus on GeoTIFFs, then I'd say `async-tiff` -> `geotiff` -> `virtual-(geo)tiff` -> `Virtualizarr`?

Which is to say, I could probably close this issue since I don't anticipate `cog3pio` having an API for determining byte ranges anymore. The ecosystem has changed significantly since this thread was opened in Apr 2024, most notably `async-tiff` arrived on the scene, and I should point you to threads like https://github.com/developmentseed/async-tiff/issues/6 if you're interested in smart range request handling.

**Comment by weiji14** [comment]
I'm gonna close this since `async-geotiff` is out and had some releases recently. Probably can continue the topic in https://github.com/developmentseed/async-geotiff/issues/78.

And just to throw this out there. I see [`async-geotiff`](https://github.com/developmentseed/async-geotiff) as implementing the reading part whereas [`geotiff`](https://github.com/georust/geotiff) would probably handle both read/write (once https://github.com/georust/geotiff/issues/16 is done). At least that's the current view based on https://github.com/developmentseed/async-tiff/issues/110#issuecomment-3050294469. While async-tiff/geotiff focuses on asynchronous fetches over the network, cog3pio's long-term plan is to handle hardware accelerated decompression and parallel decoding of the bytes in compressed tiles (see my [FOSS4G talk](https://weiji14.github.io/foss4g2025/4)).

**Comment by TomNicholas** [comment]
> cog3pio's long-term plan is to handle hardware accelerated decompression and parallel decoding of the bytes in compressed tiles

You wanna just put this in zarr? 😀

**Comment by weiji14** [comment]
> You wanna just put this in zarr? 😀

It's kinda already [there](https://docs.rapids.ai/api/kvikio/stable/zarr/)! But kvikIO is a bit too high-level of a C++ lib and gonna [undergo some architectural changes](https://github.com/apache/opendal/issues/5090#issuecomment-3501046702), so I don't want to build on top of that, and am looking at https://docs.nvidia.com/cuda/nvtiff/index.html or potentially wrapping the low-level cuFile/nvCOMP/etc instead 🙂 I'll probably keep an eye on https://github.com/zarrs/zarrs-python/issues/68 and see what they do that and there will be some shared parts.
